<center> <img src = https://raw.githubusercontent.com/AndreyRysistov/DatasetsForPandas/main/hh%20label.jpg alt="drawing" style="width:400px;">

# <center> Проект: Анализ вакансий из HeadHunter
   

In [256]:
import pandas as pd
import psycopg2

In [257]:
# вставьте сюда параметры подключения из юнита 1. Работа с базой данных из Python
DBNAME = 'project_sql'
USER = 'skillfactory'
PASSWORD = 'cCkxxLVrDE8EbvjueeMedPKt'
HOST = '84.201.134.129'
PORT = 5432

In [258]:
connection = psycopg2.connect(
    dbname=DBNAME,
    user=USER,
    host=HOST,
    password=PASSWORD,
    port=PORT
)

# Юнит 3. Предварительный анализ данных

1. Напишите запрос, который посчитает количество вакансий в нашей базе (вакансии находятся в таблице vacancies). 

In [259]:
# текст запроса
query_3_1 = f'''select Count(id) "Количество вакансий"
                from public.vacancies       
             '''

In [260]:
df = pd.read_sql_query(query_3_1, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3851344990.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_1, connection)


Количество вакансий
0                49197

2. Напишите запрос, который посчитает количество работодателей (таблица employers). 

In [261]:
query_3_2 = f'''select  Count(id) "Количество работодателей"
                from  public.EMPLOYERS   
             '''

In [262]:
df = pd.read_sql_query(query_3_2, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3588252056.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_2, connection)


Количество работодателей
0                     23501

3. Посчитате с помощью запроса количество регионов (таблица areas).

In [263]:
query_3_3 = f'''select Count(id) "Количество регионов"
                from  public.areas  
             '''

In [264]:
df = pd.read_sql_query(query_3_3, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2018224393.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_3, connection)


Количество регионов
0                 1362

4. Посчитате с помощью запроса количество сфер деятельности в базе (таблица industries).

In [265]:
query_3_4 = f'''select Count(id) "Количество сфер деятельности"
                from  public.industries  
             '''

In [266]:
df = pd.read_sql_query(query_3_4, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2172479966.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_3_4, connection)


Количество сфер деятельности
0                           294

***

# выводы по предварительному анализу данных
В данной базе данных содержиться 49197 вакансий, которые представлены 23501 работодателями из 1362 регионов и разделены на 294 сферы деятельности.

1. Напишите запрос, который позволит узнать, сколько (cnt) вакансий в каждом регионе (area).
Отсортируйте по количеству вакансий в порядке убывания.

In [267]:
# код запроса представляет из себя строковую переменную
n = 5
query_4_1 = f'''
SELECT 
        count(v.id) cnt, --считаем вакансии
        a.name area      --выводим столбец с регионом             
FROM 
        public.vacancies v 
JOIN public.areas a on v.area_id = a.id  --присоединяем к таблице вакансии, таблицу регион
GROUP BY 2 --группируем по региону
ORDER BY 1 desc --сортируем в порядке убывания
LIMIT {n}  
'''

In [268]:
df = pd.read_sql_query(query_4_1, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\347208320.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_1, connection)


cnt             area
0  5333           Москва
1  2851  Санкт-Петербург
2  2112            Минск
3  2006      Новосибирск
4  1892           Алматы

2. Напишите запрос, чтобы определить у какого количества вакансий заполнено хотя бы одно из двух полей с зарплатой.

In [269]:
query_4_2 = f'''
SELECT 
       count(id) --считаем вакансии
FROM 
       public.vacancies
WHERE 
       salary_from IS NOT NULL or salary_to IS NOT NULL --задаем условие "не пустые"
'''

In [270]:
df = pd.read_sql_query(query_4_2, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2168503709.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_2, connection)


count
0  24073

3. Найдите средние значения для нижней и верхней границы зарплатной вилки. Округлите значения до целого.

In [271]:
query_4_3 = f''' 
SELECT 
       round(avg(salary_from)) Low, -- среднее значение для нижней границы
       round(avg(salary_to)) Up   -- среднее значение для верхней границы
FROM  public.vacancies 
'''

In [272]:
df = pd.read_sql_query(query_4_3, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2222372203.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_3, connection)


low        up
0  71065.0  110537.0

4. Напишите запрос, который выведет количество вакансий для каждого сочетания типа рабочего графика (schedule) и типа трудоустройства (employment), используемого в вакансиях. Результат отсортируйте по убыванию количества.


In [273]:
query_4_4 = f'''
SELECT
        Count(d.id),
        schedule "Тип рабочего графика",
        employment "Тип трудоустройства"
FROM
        public.vacancies d
Group by 2,3
Order by 1 desc 
'''

In [274]:
df = pd.read_sql_query(query_4_4, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2981602351.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_4, connection)


count Тип рабочего графика  Тип трудоустройства
0   35367          Полный день     Полная занятость
1    7802     Удаленная работа     Полная занятость
2    1593        Гибкий график     Полная занятость
3    1312     Удаленная работа  Частичная занятость
4     940       Сменный график     Полная занятость
5     569          Полный день           Стажировка
6     367       Вахтовый метод     Полная занятость
7     347          Полный день  Частичная занятость
8     312        Гибкий график  Частичная занятость
9     141          Полный день     Проектная работа
10    133     Удаленная работа     Проектная работа
11    116        Гибкий график           Стажировка
12    101       Сменный график  Частичная занятость
13     64     Удаленная работа           Стажировка
14     18        Гибкий график     Проектная работа
15     12       Сменный график           Стажировка
16      2       Вахтовый метод     Проектная работа
17      1       Сменный график     Проектная работа

5. Напишите запрос, выводящий значения поля Требуемый опыт работы (experience) в порядке возрастания количества вакансий, в которых указан данный вариант опыта. 

In [275]:
query_4_5 = f'''
SELECT
        Count(d.id),
        experience "Требуемый опыт работы"
         
FROM
         public.vacancies d
Group by 2
Order by 1
'''

In [276]:
df = pd.read_sql_query(query_4_5, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2601166264.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_4_5, connection)


count Требуемый опыт работы
0   1337           Более 6 лет
1   7197             Нет опыта
2  14511         От 3 до 6 лет
3  26152    От 1 года до 3 лет

***

# выводы по предварительному анализу данных
Исходя из полученных результатов можно сказать, что по количеству размещенных вакансий лидируют - мегаполисы с развитой  инфраструктурой, технологическими, экономическими и промышленными кластерами, которым требуется большое количество персонала. Так же мы видим нижнюю границу зарплаты, которая состовляет 71065 и верхнюю границу зарплаты в размере 110537. Самыми востребованные работники являються кандидаты на полный рабочий день с полной занятностью с опытом работы от 1 до 3 лет. Так эже у большей половины вакансий не указана зарплата.

# Юнит 5. Анализ работодателей

1. Напишите запрос, который позволит узнать, какие работодатели находятся на первом и пятом месте по количеству вакансий.

In [277]:
query_5_1 = f'''
SELECT
       e.name "Название работодателя", 
       Count(v.id) "Количество вакансий"
FROM
       public.employers e
JOIN public.vacancies v on v.employer_id = e.id
GROUP BY 1
ORDER BY 2 desc
LIMIT 5
''' 

In [278]:
df = pd.read_sql_query(query_5_1, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\1792814825.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_1, connection)


Название работодателя  Количество вакансий
0                Яндекс                 1933
1            Ростелеком                  491
2              Тинькофф                  444
3                  СБЕР                  428
4         Газпром нефть                  331

2. Напишите запрос, который для каждого региона выведет количество работодателей и вакансий в нём.
Среди регионов, в которых нет вакансий, найдите тот, в котором наибольшее количество работодателей.


In [279]:
query_5_2 = f'''
SELECT
        a.name "Название региона",
        count(e.id) "Количество работодателей"                   
                     
FROM 
        public.areas a
LEFT JOIN public.employers e on a.id = e.area
LEFT JOIN public.vacancies v on a.id = v.area_id
GROUP BY 1 
HAVING count(v.id) = 0 --регионы где нет вакансий 
ORDER BY 2 desc -- сортируем по убыванию работодателей  
'''

In [280]:
df = pd.read_sql_query(query_5_2, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3341481329.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_2, connection)


Название региона  Количество работодателей
0                          Россия                       410
1                       Казахстан                       207
2              Московская область                        75
3              Краснодарский край                        19
4              Ростовская область                        18
..                            ...                       ...
588                      Струнино                         0
589    Славгород (Алтайский край)                         0
590                     Сорочинск                         0
591  Ленинский (Тульская область)                         0
592                         Тавда                         0

[593 rows x 2 columns]

3. Для каждого работодателя посчитайте количество регионов, в которых он публикует свои вакансии. Отсортируйте результат по убыванию количества.


In [281]:
query_5_3 = f'''
SELECT
        e.name "Название работодателя",
        count(distinct v.area_id) "Количество регионов" --считаем только уникальные значения
FROM 
        public.vacancies v
LEFT JOIN public.employers e on v.employer_id = e.id 
GROUP BY 1  
ORDER BY 2 desc -- сортируем по убыванию  
'''

In [282]:
df = pd.read_sql_query(query_5_3, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\2398953733.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_3, connection)


Название работодателя  Количество регионов
0                      Яндекс                  181
1                  Ростелеком                  152
2                  Спецремонт                  116
3      Поляков Денис Иванович                   88
4                   ООО ЕФИН                    71
...                       ...                  ...
14761                  UniSol                    1
14762            UNISTORY LLC                    1
14763                   UNIT6                    1
14764     United Distribution                    1
14765              UNITEL LLC                    1

[14766 rows x 2 columns]

4. Напишите запрос для подсчёта количества работодателей, у которых не указана сфера деятельности. 

In [283]:
query_5_4 = f'''
SELECT 
        count(e.name) "Количество работодателей"           
FROM 
        public.employers_industries ei
RIGHT JOIN public.employers e on ei.employer_id = e.id
WHERE industry_id is null
'''

In [284]:
df = pd.read_sql_query(query_5_4, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3114411492.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_4, connection)


Количество работодателей
0                      8419

5. Напишите запрос, чтобы узнать название компании, находящейся на третьем месте в алфавитном списке (по названию) компаний, у которых указано четыре сферы деятельности. 

In [285]:
query_5_5 = f'''
SELECT
        e.name "Название компании",
        count(ie.industry_id)            
                    
FROM 
        public.employers_industries ie
JOIN public.employers e on ie.employer_id = e.id
GROUP BY 1
HAVING count(ie.industry_id) = 4  -- задаем условие показать, где есть 4 сферы деятельности
ORDER BY 1
LIMIT 3 
'''

In [286]:
df = pd.read_sql_query(query_5_5, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\1650419958.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_5, connection)


Название компании  count
0      101 Интернет      4
1          21vek.by      4
2              2ГИС      4

6. С помощью запроса выясните, у какого количества работодателей в качестве сферы деятельности указана Разработка программного обеспечения.


In [287]:
query_5_6 = f'''
SELECT
       Count(employer_id) "Количество работодателей" --считаем количество работодателей            
FROM
       public.employers_industries ie
JOIN public.industries i on i.id = ie.industry_id

WHERE i.name = 'Разработка программного обеспечения' -- задаем условие
'''

In [288]:
df = pd.read_sql_query(query_5_6, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\1169039133.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_5_6, connection)


Количество работодателей
0                      3553

7. Для компании «Яндекс» выведите список регионов-миллионников, в которых представлены вакансии компании, вместе с количеством вакансий в этих регионах. Также добавьте строку Total с общим количеством вакансий компании. Результат отсортируйте по возрастанию количества.

Список городов-милионников надо взять [отсюда](https://ru.wikipedia.org/wiki/%D0%93%D0%BE%D1%80%D0%BE%D0%B4%D0%B0-%D0%BC%D0%B8%D0%BB%D0%BB%D0%B8%D0%BE%D0%BD%D0%B5%D1%80%D1%8B_%D0%A0%D0%BE%D1%81%D1%81%D0%B8%D0%B8). 

Если возникнут трудности с этим задание посмотрите материалы модуля  PYTHON-17. Как получать данные из веб-источников и API. 

In [289]:
# код для получения списка городов-милионников

In [290]:
# текст запроса

In [291]:
# результат запроса

***

# выводы по предварительному анализу данных
Лидером по количеству вакансий, представленных компанией однозначно являеться "Яндекс" со своими 1933 вакансиями. Компании Ростелеком, Тиньков и Сбер представленны почти одинаковым количеством вакансии от 428 до 491.
Дальше мы пытались получить данные по работодателям и их вакнсиям в регионах и тут появились не совсем корректные данные,  в виду не четкого определения региона. Если смотреть на эти данные то в России представлено больше 410 работодателей, а в Казахстане всего 207. Если же мы рассматриваем регион, только на территории России, то можно отметить, что в большинстве регионов не представленно ни одного работадателя, а лидерами являються Московская область - 75, Краснодарский край - 19, Ростовская область - 18. 
Яндекс	181, Ростелеком	152 - это наиболее крубные работодатели и количество регионов в которых представлены их вакансии. Обе эти компании связаны с IT, связью, интернетом, поэтому имеют свои представительства во многих регионах. 
Интересно что у 8419 компаний вообще не указана сфера деятельности.  
3553 компаний занимаются разработкой ПО. 

# Юнит 6. Предметный анализ

1. Сколько вакансий имеет отношение к данным?

Считаем, что вакансия имеет отношение к данным, если в её названии содержатся слова 'data' или 'данн'.

*Подсказка: Обратите внимание, что названия вакансий могут быть написаны в любом регистре.* 


In [293]:
query_6_1 = f'''
SELECT
       count(v.name) as "data вакансии" 
FROM 
       public.vacancies v 
WHERE v.name ILIKE '%data%' OR v.name ILIKE '%данн%'
'''

In [294]:
df = pd.read_sql_query(query_6_1, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\485020324.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_1, connection)


data вакансии
0           1771

2. Сколько есть подходящих вакансий для начинающего дата-сайентиста? 
Будем считать вакансиями для дата-сайентистов такие, в названии которых есть хотя бы одно из следующих сочетаний:
* 'data scientist'
* 'data science'
* 'исследователь данных'
* 'ML' (здесь не нужно брать вакансии по HTML)
* 'machine learning'
* 'машинн%обучен%'

** В следующих заданиях мы продолжим работать с вакансиями по этому условию.*

Считаем вакансиями для специалистов уровня Junior следующие:
* в названии есть слово 'junior' *или*
* требуемый опыт — Нет опыта *или*
* тип трудоустройства — Стажировка.
 

In [295]:
query_6_2 = f'''
SELECT
      count(v.id) as "Вакансии дата-сайентистов"
                       
FROM
      public.vacancies v
WHERE ((v.name) ilike '%data scientist%' or (v.name) ilike '%data science%' or
        (v.name) ilike '%исследователь данных%' or (v.name like '%ML%' and v.name not like '%HTML%') or
        (v.name) ilike '%machine learning%' or (v.name) ilike '%машинн%обучен%%') and
        (v.experience like '%Нет опыта%' or v.experience like '%Стажировка%'or (v.name) ilike '%junior%') 
'''

In [296]:
df = pd.read_sql_query(query_6_2, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3641650667.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_2, connection)


Вакансии дата-сайентистов
0                         51

3. Сколько есть вакансий для DS, в которых в качестве ключевого навыка указан SQL или postgres?

** Критерии для отнесения вакансии к DS указаны в предыдущем задании.*

In [297]:
query_6_3 = f'''
SELECT
        count(v.id) as "Вакансии с навыками SQL или postgres"
FROM 
        public.vacancies v
WHERE
     ((v.name) ilike '%data scientist%' or (v.name) ilike '%data science%' or
       (v.name) ilike '%исследователь данных%' or (v.name like '%ML%' and v.name not like '%HTML%') or
        (v.name) ilike '%machine learning%' or (v.name) ilike '%машинн%обучен%%')  and 
        (v.key_skills like '%SQL%' or v.key_skills like '%postgres%')
'''

In [298]:
df = pd.read_sql_query(query_6_3, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\1159587760.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_3, connection)


Вакансии с навыками SQL или postgres
0                                   201

4. Проверьте, насколько популярен Python в требованиях работодателей к DS.Для этого вычислите количество вакансий, в которых в качестве ключевого навыка указан Python.

** Это можно сделать помощью запроса, аналогичного предыдущему.*

In [299]:
query_6_4 = f'''
SELECT
        count(v.id) as "Вакансии с ключевым навыком Python"
FROM 
        public.vacancies v
WHERE
     ((v.name) ilike '%data scientist%' or (v.name) ilike '%data science%' or
       (v.name) ilike '%исследователь данных%' or (v.name like '%ML%' and v.name not like '%HTML%') or
        (v.name) ilike '%machine learning%' or (v.name) ilike '%машинн%обучен%%')  and 
        (v.key_skills like '%Python%')
'''

In [300]:
df = pd.read_sql_query(query_6_4, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\972386973.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_4, connection)


Вакансии с ключевым навыком Python
0                                 351

5. Сколько ключевых навыков в среднем указывают в вакансиях для DS?
Ответ округлите до двух знаков после точки-разделителя.

In [301]:
query_6_5 = f'''
SELECT
        round(avg(LENGTH(v.key_skills) - LENGTH(replace(v.key_skills, chr(9), '')) + 1), 2) "Среднее значение навыков"
FROM 
        public.vacancies v
WHERE
     ((v.name) ilike '%data scientist%' or (v.name) ilike '%data science%' or
       (v.name) ilike '%исследователь данных%' or (v.name like '%ML%' and v.name not like '%HTML%') or
        (v.name) ilike '%machine learning%' or (v.name) ilike '%машинн%обучен%%') 
'''

In [302]:
df = pd.read_sql_query(query_6_5, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\390248827.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_5, connection)


Среднее значение навыков
0                      6.41

6. Напишите запрос, позволяющий вычислить, какую зарплату для DS в **среднем** указывают для каждого типа требуемого опыта (уникальное значение из поля *experience*). 

При решении задачи примите во внимание следующее:
1. Рассматриваем только вакансии, у которых заполнено хотя бы одно из двух полей с зарплатой.
2. Если заполнены оба поля с зарплатой, то считаем зарплату по каждой вакансии как сумму двух полей, делённую на 2. Если заполнено только одно из полей, то его и считаем зарплатой по вакансии.
3. Если в расчётах участвует null, в результате он тоже даст null (посмотрите, что возвращает запрос select 1 + null). Чтобы избежать этой ситуацию, мы воспользуемся функцией [coalesce](https://postgrespro.ru/docs/postgresql/9.5/functions-conditional#functions-coalesce-nvl-ifnull), которая заменит null на значение, которое мы передадим. Например, посмотрите, что возвращает запрос `select 1 + coalesce(null, 0)`

Выясните, на какую зарплату в среднем может рассчитывать дата-сайентист с опытом работы от 3 до 6 лет. Результат округлите до целого числа. 

In [303]:
query_6_6 = f'''
SELECT
       round(avg(coalesce((salary_from+salary_to)/2, salary_from, salary_to))) --"Средняя зарплата с опытом 3-6 лет"
FROM 
        public.vacancies v
WHERE
     ((v.name) ilike '%data scientist%' or (v.name) ilike '%data science%' or
       (v.name) ilike '%исследователь данных%' or (v.name like '%ML%' and v.name not like '%HTML%') or
        (v.name) ilike '%machine learning%' or (v.name) ilike '%машинн%обучен%%')and
        (v.salary_from is not null or salary_to is not null) and 
        (v.experience = 'От 3 до 6 лет')
'''

In [304]:
df = pd.read_sql_query(query_6_6, connection)
df

C:\Users\Alla\AppData\Local\Temp\ipykernel_11296\3561045109.py:1: UserWarning: pandas only supports SQLAlchemy connectable (engine/connection) or database string URI or sqlite3 DBAPI2 connection. Other DBAPI2 objects are not tested. Please consider using SQLAlchemy.
  df = pd.read_sql_query(query_6_6, connection)


round
0  243115.0

# выводы по предварительному анализу данных
Исходя из проведенного анализа для вакансий связанных с данными можно сказть, что таких насчитывается 1771, однако из них только 51 подходит junior. Вакансия для дата-саентистов с  ключевыми навыка SQL или postgres всего 201, а для ключевого навака знания python 351. В всреднем для дата-саентистов в вакансии указывают более 6 ключевых навыков, а для кандидатов с опытом работы 3-6 лет можено претендовать на зарплату в 243115 руб.

***

# Общий вывод по проекту

In [306]:
# подведем итог исследования, обобщите выводы
# здесь можно (это будет плюсом) провести дополнительные исследования данных, сделать прогнозы, продумать варианты продолжения исследования

В выборке представлены 49197 вакансий от 23501 работодателей в 1362 регионах по 294 сферам деятельности.

Лидерами среди регионов по количеству размещенных вакансий являются - мегаполисы с развитой инфраструктурой.

Среднее значение по нижней границе зарплаты - 71065, по верхней границе ЗП - 110537.
Наиболее востребованные работники - работники на полный рабочий день с полной занятностью с опытом работы от 1 до 3 лет.

По количеству вакансий лидирует компания "Яндекс" - 1933 вакансии. Компании Ростелеком, Тиньков и Сбер представленны почти одинаковым количеством вакансии от 428 до 491. Яндекс	181, Ростелеком	152 - это наиболее крубные работодатели и количество регионов в которых представлены их вакансии. Обе эти компании связаны с IT, связью, интернетом, поэтому имеют свои представительства во многих регионах. 
У 8419 компаний вообще не указана сфера деятельности, а 3553 компаний занимаются разработкой ПО.
Вакансий связанных с данными  1771, из них только 51 подходит junior. Вакансия для дата-саентистов с  ключевыми навыка SQL или postgres всего 201, а для ключевого навака знания python 351. В всреднем для дата-саентистов в вакансии указывают более 6 ключевых навыков, а для кандидатов с опытом работы 3-6 лет можено претендовать на зарплату в 243115 руб.